text/3634-scoped-impl-trait-for-type.md

```diff
@@ -0,0 +1,3436 @@
+
+[`Hash`]: https://doc.rust-lang.org/stable/std/hash/trait.Hash.html
+[`PartialEq`]: https://doc.rust-lang.org/stable/std/cmp/trait.PartialEq.html
+[`Eq`]: https://doc.rust-lang.org/stable/std/cmp/trait.Eq.html
+[`PartialOrd`]: https://doc.rust-lang.org/stable/std/cmp/trait.PartialOrd.html
+[`Ord`]: https://doc.rust-lang.org/stable/std/cmp/trait.Ord.html
+
+[`Deserialize`]: https://docs.rs/serde/1/serde/trait.Deserialize.html
+[`Serialize`]: https://docs.rs/serde/1/serde/trait.Serialize.html
+
+[the newtype pattern]: https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-the-newtype-pattern-to-implement-external-traits-on-external-types
+
+### Publishing and Importing Scoped Implementations
+
+You can also publish a scoped implementation further by adding a visibility before `use` ...:
+
+```rust
+pub use impl Trait for Type {
+    // ...
+}
+
+pub use unsafe impl UnsafeTrait for Type {
+    // ...
+}
+```
+
+... and import it into other scopes:
+
+```rust
+use other_module::{
+    impl Trait for Type,
+    impl UnsafeTrait for Type,
+};
+```
+
+Note that the scoped implementation of `UnsafeTrait` is imported without the `unsafe` keyword. **It is the implementing crate's responsibility to ensure the exported `unsafe` implementation is sound everywhere it is visible!**
+
+Generic parameters, bounds and `where`-clauses can be used as normal in each of these locations, though you usually have to brace `impl</*...*/> Trait for Type where /*...*/` individually in `use`-declarations.
+
+You can import a subset of a generic implementation, by narrowing bounds or replacing type parameters with concrete types in the `use`-declaration.
+
+Global implementations can be imported from the root namespace, for example to shadow a scoped implementation:
+
+```rust
+use ::{impl Trait for Type};
+```
+
+### Scoped implementations and generics
+[scoped-implementations-and-generics]: #scoped-implementations-and-generics
+
+Scoped implementations are resolved on most generics' type arguments where those are specified, and become part of the (now less generic) host type's identity:
+
+```rust
+#[derive(Default)]
+struct Type<T>(T);
+
+trait Trait {
+    fn trait_fn();
+}
+
+impl<T: Trait> Type<T> {
+    fn type_fn() {
+        T::trait_fn();
+    }
+}
+
+mod nested {
+    use super::{Trait, Type};
+
+    use impl Trait for () {
+        fn trait_fn() {
+            println!("nested");
+        }
+    }
+
+    pub type Alias = Type<()>;
+}
+use nested::Alias;
+
+fn main() {
+    Alias::type_fn(); // "nested"
+
+    // Type::<()>::type_fn();
+    //             ^^^^^^^ error[E0599]: the function or associated item `type_fn` exists for struct `Type<()>`, but its trait bounds were not satisfied
+
+    // let t: Type<()> = Alias::default();
+    //                   ^^^^^^^^^ error[E0308]: mismatched types
+
+    let t: Type<() as Trait in nested> = Alias::default();
```

Comment:
Yes, this is a mistake.  An inline implementation environment can only refer to implementations that it can actually see.

Fixed in 129e734ae1bc198becb49b2b4590232e0a746c5e.